### PR TITLE
fix(lambda): support Lambda's new `Invoke` with `Qualifier` authorization strategy

### PR DESCRIPTION
**Pull request by madeline-k**
‼️ Lambda is changing their authorization strategy, which means that some behavior that was previously valid now results in `access-denied` errors.

Under the new behavior, customer lambda invocations will fail if the CDK generates a policy with an unqualified ARN as the resource, and the customer invokes lambda with the unqualified ARN and the `Qualifier` request parameter. 

Example of an affected setup:

```
Statement: 
{
  Effect: "Allow",
  Action: "lambda:InvokeFunction",
  Resource: "arn:aws:lambda:...:function:MyFunction",
}

API Call:
lambda.Invoke({
  FunctionName: "MyFunction",
  Qualifier: "1234",
})
```

This `Invoke` call *used* to succeed, but under the new authorization strategy it will fail. The required statement to make the call succeed would be (note the qualified ARN):

```
{
  Effect: "Allow",
  Action: "lambda:InvokeFunction",
  Resource: "arn:aws:lambda:...:function:MyFunction:1234",
}
```

This PR aims to align the CDK with the new authorization strategy. The PR introduces changes to the `grantInvoke()` api on a lambda function. Now, when calling `grantInvoke()` on a lambda function, `[ARN, ARN:*]` is used in the identity policy, so that identities that are granted permission to invoke the Function may also invoke all of its Versions and Aliases. 

When calling `grantInvoke()` on a lambda function Version or Alias, the generated identity policy will remain the same, and only include `ARN:<version/alias>` in the policy.

This is part of #19273

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by madeline-k** [comment]
Adding do not merge label until we decide whether or not to Feature Flag this change

**Comment by kaizen3031593** [comment]
If I'm not mistaken, when this is complete it should also close #17515.

**Comment by madeline-k** [comment]
> If I'm not mistaken, when this is complete it should also close #17515.

@kaizen3031593,  It will not. But I am planning to do that in another PR. This PR only adds [ARN, ARN:*] when you call `grantInvoke` on a "regular" Function. Not on a `Version` or `Alias`. 

In a separate PR, we will add the [ARN, ARN:*] permissions for calling `grantInvoke` on a `Version`, when it is used in a stepfunction task. That part will fix #17515.

On a somewhat related note - calling `grantInvoke` on an `Alias` will always remain the same, and only grant permissions to ARN:alias, because one of the purposes of an Alias is to restrict users to only have access to a specific version of the function.

File: packages/@aws-cdk/custom-resources/lib/provider-framework/waiter-state-machine.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/custom-resources/lib/provider-framework/waiter-state-machine.ts:53]: Is there a reason why we're okay with `grantInvoke()` here? Or is this legacy and we should be doing the whole

```ts
new PolicyStatement({
  actions: ['lambda:InvokeFunction'],
  resources: props.timeoutHandler.resourceArnsForGrantInvoke,
});
```

Is it because we control both the role and the lambdas so we can be certain they are in the same account?

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-lambda/lib/function-base.ts:303]: How do we feel about this? It's not the most elegant but I can't think of any other valid and sensible configuration that also catches `lambda:InvokeFunction`. I guess also `*`?

File: packages/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts:108]: This one didn't have a `// TODO` on it before but I feel like it's reasonable to change this one also. I didn't do it yet just in case it was omitted for a reason.

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-lambda/lib/function-base.ts:304]: I'm sure we will have thoughts on the warning message, so I'm kicking that conversation off here. Are we going to send a CLI notice about this change as well?

File: packages/@aws-cdk/custom-resources/lib/provider-framework/waiter-state-machine.ts
Comment (madeline-k) [reply packages/@aws-cdk/custom-resources/lib/provider-framework/waiter-state-machine.ts:53]: we can only call `grantInvoke` on something that implements `IGrantable` since we have the `role` available here, we can use `grantInvoke`. (In the step-functions-tasks cases this doesn't work because we don't have access to an `IGrantable` when the policy statements are being created.  Does that answer the question?

File: packages/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts
Comment (madeline-k) [reply packages/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts:108]: Good catch, you're right. This one needs to be updated too.

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (madeline-k) [reply packages/@aws-cdk/aws-lambda/lib/function-base.ts:303]: Can you do the change for Resource based policies in a separate PR? It will be easier to evaluate separately.

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (madeline-k) [reply packages/@aws-cdk/aws-lambda/lib/function-base.ts:304]: How about: "AWS Lambda has changed their authorization strategy, which may cause client invocations of the lambda function <ARN> to fail with Access Denied errors. If you are using a lambda Version or Alias, make sure to call grantInvoke or addPermission on the Version or Alias, and not the underlying Function. See <github>."

Is there a way to only give this warning if their app includes a  Version or Alias somewhere of this unqualified Function?

(Again, put this in a separate PR please)

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-lambda/lib/function-base.ts:567]: Do all of the Qualified functions actually use this base class? Looks like 'LatestVersion' does not.

**Review by madeline-k — COMMENTED** [review]
(I can't request changes on a PR I created, haha )

File: packages/@aws-cdk/aws-redshift/test/integ.database.expected.json
Comment (madeline-k) [inline packages/@aws-cdk/aws-redshift/test/integ.database.expected.json:1453]: This might seem hacky, but I think you should revert the Asset hash changes in this file. They are irrelevant to the change at hand. If they are out of date somehow, and this integ test needs updating, that should be a separate Pr.

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-lambda/lib/function-base.ts:69]: Why are you adding this to the interface? (Not saying it is incorrect, just want to understand why)

File: packages/@aws-cdk/custom-resources/test/provider-framework/integ.provider.expected.json
Comment (madeline-k) [inline packages/@aws-cdk/custom-resources/test/provider-framework/integ.provider.expected.json:1055]: Looks like a lot of our integration tests might be out of date. Can we update this separately? It unrelated to the lambda change.

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (madeline-k) [reply packages/@aws-cdk/aws-lambda/lib/function-base.ts:304]: And yes, I think we are planning to do a CLI notice. We should work with the Lambda team to craft the right verbiage for that.

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (kaizen3031593) [reply packages/@aws-cdk/aws-lambda/lib/function-base.ts:69]: Most/all integrations accept an `IFunction`, so we don't have access to `FunctionBase.resourceArnsForGrantInvoke` and instead need to use `IFunction.resourceArnsForGrantInvoke`. This is true for basically all of the stepfunctions-tasks lambda integrations.

File: packages/@aws-cdk/custom-resources/test/provider-framework/integ.provider.expected.json
Comment (kaizen3031593) [reply packages/@aws-cdk/custom-resources/test/provider-framework/integ.provider.expected.json:1055]: It is unrelated. I'll open a PR to address these separately.

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (kaizen3031593) [reply packages/@aws-cdk/aws-lambda/lib/function-base.ts:567]: Did some research on this, but will update it on the new PR i've created.

File: packages/@aws-cdk/custom-resources/test/provider-framework/integ.provider.expected.json
Comment (kaizen3031593) [reply packages/@aws-cdk/custom-resources/test/provider-framework/integ.provider.expected.json]: I ran `yarn integ --dry-run` on all integ tests, and this is what i got. i don't think there's a way for the integ tests to be out of date, as they get checked with `expected.json` on every build. So these integ changes have something to do with the current PR.

File: packages/@aws-cdk/aws-redshift/test/integ.database.expected.json
Comment (kaizen3031593) [reply packages/@aws-cdk/aws-redshift/test/integ.database.expected.json]: I think the outdated integ tests are fixed after the latest merge from master

**Comment by kaizen3031593** [comment]
Adding `pr-linter/exempt-readme` and `pr-linter/exempt-integ-test` for now but we can discuss whether either exemption is appropriate later

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-lambda/lib/function-base.ts:67]: Let's add some extra detail to this comment that users should not use this if they can avoid it, and should use `grantInvoke` instead

File: packages/@aws-cdk/aws-redshift/test/integ.database.expected.json
Comment (madeline-k) [inline packages/@aws-cdk/aws-redshift/test/integ.database.expected.json:1183]: Unrelated integ test change here?

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/test/integ.s3-bucket.lit.expected.json
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/test/integ.s3-bucket.lit.expected.json:616]: This is weird. This permission looks to be duplicated now?

**Review by madeline-k — COMMENTED** [review]
I think the code pretty much looks good now (Although, the build is failing?). 

Now we need to make sure we clearly explain what is going on here in three places.

1. API ref (comment)
2. README
3. The Changelog (PR Description). 

2 could be separated, but 1&3 need to be addressed in this PR. @rix0rrr, any thoughts here? I am not sure the best way to communicate this.

File: packages/@aws-cdk/aws-redshift/test/integ.database.expected.json
Comment (kaizen3031593) [reply packages/@aws-cdk/aws-redshift/test/integ.database.expected.json]: updated the integ test for the millionth time. No idea why this keeps getting out of sync.

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/test/integ.s3-bucket.lit.expected.json
Comment (kaizen3031593) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/test/integ.s3-bucket.lit.expected.json]: Sigh.... I updated them again, and now the assets have changed. I guess the source of truth is what the PR build thinks the integ test should be, so i will work towards getting a successful build.

**Comment by kaizen3031593** [comment]
I updated 1. API ref, but for 2 & 3 I am going to need some help concocting a coherent message. I'll take a stab at 3. and update the description to my liking.

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-lambda/lib/function-base.ts:69]: Nice 👍

**Review by madeline-k — COMMENTED** [review]
Really like your updates to the PR description 👍 

Thanks for fighting with the integ tests! Let me know if I can help

**Comment by kaizen3031593** [comment]
Finally got the build working for this one. Some of the integ tests have changed assets, at this point, I think that we should be okay with it. The build fails without these changes.

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**
